### PR TITLE
ref(ui): Enhance Carousel component

### PR DESCRIPTION
**Pull request by evanpurkhiser**
1. Record a visibility list and use that to determine if we need to scroll left or right. This has the advantage setting isAtStart and isAtEnd as state in that we're able to be a little smarter about if an element is visibile or not.

    In addition here we've added a `visibleRatio` prop to the component, which lets you customize this value.

 2. Use scrollIntoView. This is more robust and allows us to set a `scroll-margin` on the children.

 3. Update tests to test for button clicks

### CODE REVIEW

File: static/app/components/carousel.spec.tsx [hunk 1]
```diff
@@ -1,79 +1,140 @@
...
+    takeRecords = jest.fn();
 
+    constructor(cb: IntersectionObserverCallback) {
+      // @ts-expect-error The callback wants just way too much stuff for our simple mock
+      intersectionOnbserverCb = cb;
+    }
+    observe() {}
+    unobserve() {}
+    disconnect() {}
+  };
```
Comment (vuluongj20) [inline static/app/components/carousel.spec.tsx:22]: Do we still need this if it's in `tests/js/setup`?

Comment (evanpurkhiser) [reply static/app/components/carousel.spec.tsx:22]: yeah since I need to override the callback

File: static/app/components/carousel.tsx [hunk 2]
```diff
@@ -1,92 +1,115 @@
...
+            // This is why we fallback to the currentThreshold value if there
+            // was no entry for the child.
+            return entry !== undefined
+              ? entry.intersectionRatio > visibleRatio
+              : currentVisibility[idx] ?? false;
+          })
+        ),
+      {
+        root: ref.current,
+        threshold: [0, visibleRatio, 1],
```
Comment (vuluongj20) [inline static/app/components/carousel.tsx:66]: hmm why do we need the 0 & 1 thresholds? looks like we only need to update visibility when the interaction ratio crosses `visibleRatio`, so shouldn't `[visibleRatio]` suffice?

Comment (evanpurkhiser) [reply static/app/components/carousel.tsx:66]: oh good point 👍

Comment (evanpurkhiser) [reply static/app/components/carousel.tsx:66]: Fixed

**Review by vuluongj20 — APPROVED** [review]

**Pull request merged**